File: features/rolemanagement/src/main/java/com/nivasafinance/features/rolemanagement/aspect/AuthorizationAspect.java
package com.nivasafinance.features.rolemanagement.aspect;

import com.nivasafinance.common.context.UserContext;
import com.nivasafinance.common.exception.ForbiddenException;
import com.nivasafinance.common.exception.UnauthorizedException;
import com.nivasafinance.common.annotations.RequirePermission;
import com.nivasafinance.features.rolemanagement.permissionchecker.PermissionCheckerService;
import lombok.RequiredArgsConstructor;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Component;

@Aspect
@Component
@Order(1)
@RequiredArgsConstructor
public class AuthorizationAspect {
    
    private static final Logger logger = LoggerFactory.getLogger(AuthorizationAspect.class);
    
    private final PermissionCheckerService permissionCheckerService;
    
    @Around("@annotation(requirePermission)")
    public Object checkPermission(ProceedingJoinPoint joinPoint, RequirePermission requirePermission) throws Throwable {
        String userName = UserContext.getUsername();
        if (userName == null) {
            throw new UnauthorizedException("User not authenticated");
        }
        
        String permissionName = requirePermission.permissionName();
        boolean hasPermission = permissionCheckerService.checkPermissionForUser(
                userName,
                permissionName
        );
        
        if (!hasPermission) {
            logger.warn(
                    "Access denied for user {} to permission {}",
                    userName,
                    permissionName
            );
            throw new ForbiddenException(
                    "Insufficient permissions: " + permissionName
            );
        }
        
        logger.debug(
                "Permission granted for user {} to permission {}",
                userName,
                permissionName
        );
        return joinPoint.proceed();
    }
}

File: features/rolemanagement/src/main/java/com/nivasafinance/features/rolemanagement/aspect/AuthorizationAspect.kt
package com.nivasafinance.features.rolemanagement.aspect

import com.nivasafinance.common.exception.ForbiddenException
import com.nivasafinance.common.exception.UnauthorizedException
import com.nivasafinance.features.rolemanagement.annotation.RequirePermission
import com.nivasafinance.features.rolemanagement.permissionchecker.PermissionCheckerService
import com.nivasafinance.security.context.UserContext
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.slf4j.LoggerFactory
import org.springframework.core.annotation.Order
import org.springframework.stereotype.Component

@Aspect
@Component
@Order(1)
class AuthorizationAspect(
    private val permissionCheckerService: PermissionCheckerService
) {

    private val logger = LoggerFactory.getLogger(AuthorizationAspect::class.java)

    @Around("@annotation(requirePermission)")
    fun checkPermission(joinPoint: ProceedingJoinPoint, requirePermission: RequirePermission): Any? {
        val userInfo = UserContext.getUserInfo()
            ?: throw UnauthorizedException("User not authenticated")

        // Check permission using the existing permission checker service
        val hasPermission = permissionCheckerService.checkPermissionForUser(
            userInfo.username,
            requirePermission.action,
            requirePermission.module,
            requirePermission.operation
        )

        if (!hasPermission) {
            logger.warn(
                "Access denied for user ${userInfo.username} to ${requirePermission.action} ${requirePermission.module}"
            )
            throw ForbiddenException(
                "Insufficient permissions: ${requirePermission.action} on ${requirePermission.module}"
            )
        }

        logger.debug(
            "Permission granted for user ${userInfo.username} to ${requirePermission.action} ${requirePermission.module}"
        )
        return joinPoint.proceed()
    }
}
